cpp/oneapi/dal/backend/primitives/reduction/reduction_rm_rw_dpc.cpp

```diff
@@ -0,0 +1,395 @@
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/backend/primitives/reduction/reduction_rm_rw_dpc.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+inline auto max_wg(const sycl::queue& q) {
+    return q.get_device().template get_info<sycl::info::device::max_work_group_size>();
+}
+
+inline std::int64_t local_mem_size(const sycl::queue& q) {
+    return q.get_device().template get_info<sycl::info::device::local_mem_size>();
+}
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+class kernel_reduction_rm_rw_wide {
+    using inp_t = const Float*;
+    using out_t = Float*;
+
+public:
+    kernel_reduction_rm_rw_wide(inp_t const input_,
+                                out_t const output_,
+                                const std::int32_t width_,
+                                const std::int32_t lstride_,
+                                const BinaryOp& binary_,
+                                const UnaryOp& unary_)
+            : input{ input_ },
+              output{ output_ },
+              unary{ unary_ },
+              binary{ binary_ },
+              width{ width_ },
+              lstride{ lstride_ } {}
+
+    void operator()(sycl::nd_item<2> it) const {
+        using sycl::ONEAPI::reduce;
+        // Common for whole WG
+        const auto row_idx = it.get_global_id(0);
+        const auto loc_idx = it.get_global_id(1);
+        const auto range = it.get_global_range(1);
+        // It should be converted to upper type by default
+        inp_t const inp_row = input + lstride * row_idx;
+        // Exclusive for EU
+        Float acc = binary.init_value;
+        for (std::int32_t i = loc_idx; i < width; i += range) {
+            acc = binary.native(acc, unary(inp_row[i]));
+        }
+        // WG reduction
+        auto grp = it.get_group();
+        output[row_idx] = reduce(grp, acc, binary.native);
+    }
+
+private:
+    inp_t const input;
+    out_t const output;
+    const UnaryOp unary;
+    const BinaryOp binary;
+    const std::int32_t width;
+    const std::int32_t lstride;
+};
+
+template <typename Float, typename BinaryOp, typename UnaryOp>
+reduction_rm_rw_wide<Float, BinaryOp, UnaryOp>::reduction_rm_rw_wide(sycl::queue& q_,
```

Comment:
Code style `q_`. Underscore used only for class/struct private variables. 
Please apply to other function parameters and class/struct scope variables.

Reply:
Fixed